### PR TITLE
Add VCD export

### PR DESCRIPTION
**Pull request by nakengelhardt**
The third in a series of PRs trying to cut #4 into somewhat more manageable chunks. [Click here](https://github.com/nakengelhardt/btor2tools/compare/add_arrays...nakengelhardt:add_vcd) to view only the changes introduced in this PR (after #7).

This introduces the class `BtorSimVCDWriter` that does nearly all handling of recording the value changes. At the end of `simulate_step`, all the state values are passed to `vcd_writer->add_value_change`. The VCD writer keeps track of previous values and compares them to decide if a value has changed. It keeps the list of past value changes in memory (as a list of vcd-formatted strings) and only writes them out after the simulation finishes, because memory cells that are read or written get added to the VCD, and these locations are not known a priori, which means that the VCD header may grow during execution.

New command line arguments introduced in this PR:
The `--vcd` option enables VCD output.
The `--hierarchical-symbols` flag lets the VCD writer infer module hierarchy from dot notation in the symbols. (The previous special treatment of '$' has been removed, this is now the only thing it does.)
The `--info` option allows specifying a file that contains additional information for the VCD file. The format is whitespace separated key-value, one line per pair. Key values are:
- `name <string>`: sets the name of the top module in the VCD.
- `(posedge|negedge|event) <int>`: treats the state with the given ID as a clock of the given polarity, in the first two cases introducing value changes inbetween simulation steps, and for `event` changing the type of VCD signal.
- `bad <int>`: treat the state with the given ID as a bad state and check that it stays false during simulation. (For use in creating cover traces, as discussed [here](https://github.com/Boolector/btor2tools/pull/4#discussion_r405573892))

### CODE REVIEW

File: src/btorsim/btorsim.cpp [hunk 1]
```diff
@@ -910,6 +884,30 @@ simulate_step (int64_t k, int32_t randomize_states_that_are_inputs)
...
+      // these are not signals
+      if (l->tag == BTOR2_TAG_sort || l->tag == BTOR2_TAG_init
+          || l->tag == BTOR2_TAG_next || l->tag == BTOR2_TAG_bad
+          || l->tag == BTOR2_TAG_constraint || l->tag == BTOR2_TAG_fair
+          || l->tag == BTOR2_TAG_justice)
+        continue;
+      // only add named signals
+      if (!l->symbol) continue;
+      vcd_writer->add_value_change (k, i, current_state[i]);
+    }
```
Comment (mpreiner) [inline src/btorsim/btorsim.cpp:910]:
```suggestion
  if (dump_vcd)
  {
    for (int i = 0; i < num_format_lines; i++)
    {
      Btor2Line *l = btor2parser_get_line_by_id (model, i);
      if (!l) continue;
      // these are not signals
      if (l->tag == BTOR2_TAG_sort || l->tag == BTOR2_TAG_init
          || l->tag == BTOR2_TAG_next || l->tag == BTOR2_TAG_bad
          || l->tag == BTOR2_TAG_constraint || l->tag == BTOR2_TAG_fair
          || l->tag == BTOR2_TAG_justice)
        continue;
      // only add named signals
      if (!l->symbol) continue;
      vcd_writer->add_value_change (k, i, current_state[i]);
    }
  }
```

File: src/btorsim/btorsim.cpp [hunk 2]
```diff
@@ -910,6 +884,30 @@ simulate_step (int64_t k, int32_t randomize_states_that_are_inputs)
              (int64_t) bads.size());
     }
   }
+
+  for (auto it: extra_constraints)
```
Comment (mpreiner) [inline src/btorsim/btorsim.cpp:888]: Please add a comment about what this loop does.

File: src/btorsim/btorsim.cpp [hunk 3]
```diff
@@ -1466,8 +1464,9 @@ parse_sat_witness ()
   int64_t k = 0;
   while (parse_frame (k)) k++;
 
-  if (!found_initial_frame && states.size() > 0) parse_error ("initial frame missing");
+  // if (!found_initial_frame && states.size() > 0) parse_error ("initial frame missing");
```
Comment (mpreiner) [inline src/btorsim/btorsim.cpp:1467]: Remove?

File: src/btorsim/btorsim.cpp [hunk 4]
```diff
@@ -1492,7 +1491,7 @@ parse_unknown_witness ()
 
   while (parse_frame (k)) k++;
 
-  if (!found_initial_frame && states.size() > 0) parse_error ("initial frame missing");
+  // if (!found_initial_frame && states.size() > 0) parse_error ("initial frame missing");
```
Comment (mpreiner) [inline src/btorsim/btorsim.cpp:1494]: Remove?

File: src/btorsim/btorsimvcd.cpp [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+      vcd_file << "$var " << type << " " << width << " " << bv_identifiers[i.first] << " " << symbol << " $end\n";
+    }
+    else
+      for (auto j : am_identifiers)
+        if(j.first.first == id)
+        {
+          int64_t idx = j.first.second;
+          std::string am_ident = j.second;
+          vcd_file << "$var wire " << width << " " << am_ident << " " << symbol << "<" << std::hex << idx << std::dec << "> $end\n";
+        }
```
Comment (mpreiner) [inline src/btorsim/btorsimvcd.cpp:206]:
```suggestion
    else
      for (auto j : am_identifiers)
      {
        if(j.first.first == id)
        {
          int64_t idx = j.first.second;
          std::string am_ident = j.second;
          vcd_file << "$var wire " << width << " " << am_ident << " " << symbol << "<" << std::hex << idx << std::dec << "> $end\n";
        }
      }
```

File: src/btorsim/btorsimvcd.cpp [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+        update_time(k);
+        for (auto it: state.array_state->data)
+          if (!prev_value[id].array_state || prev_value[id].array_state->data.find(it.first)==prev_value[id].array_state->data.end() || prev_value[id].array_state->data.at(it.first) != it.second)
+          {
+            std::string sval("");
+            if (it.second->width > 1 ) sval += "b";
+            sval += btorsim_bv_to_string(it.second);
+            if (it.second->width > 1 ) sval += " ";
+            value_changes.push_back(sval + get_am_identifier(id, it.first));
+          }
```
Comment (mpreiner) [inline src/btorsim/btorsimvcd.cpp:312]:
```suggestion
        for (auto it: state.array_state->data)
        {
          if (!prev_value[id].array_state || prev_value[id].array_state->data.find(it.first)==prev_value[id].array_state->data.end() || prev_value[id].array_state->data.at(it.first) != it.second)
          {
            std::string sval("");
            if (it.second->width > 1 ) sval += "b";
            sval += btorsim_bv_to_string(it.second);
            if (it.second->width > 1 ) sval += " ";
            value_changes.push_back(sval + get_am_identifier(id, it.first));
          }
        }
```

File: src/btorsim/btorsimvcd.h [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+  void update_time (int64_t k);
+  void add_value_change (int64_t k, int64_t id, BtorSimState state);
+  ModuleTreeNode* sort_names (Btor2Parser *model, std::string topname);
+  void write_node_header (ModuleTreeNode* top);
+  std::map<int64_t, std::string> read_info_file(const char* info_path);
+  BtorSimVCDWriter (const char* vcd_path, bool readable_vcd, bool symbol_fmt);
+  ~BtorSimVCDWriter ();
+};
+
+#endif
```
Comment (mpreiner) [inline src/btorsim/btorsimvcd.h:69]: Please document header.

**Comment by mpreiner** [comment]
Also again, please clang-format all changes.

File: src/btorsim/btorsim.cpp [hunk 8]
```diff
@@ -74,7 +49,12 @@ static const char *usage =
     "  -b <n>    fake simulation to satisfy bad state property 'b<n>'\n"
     "  -j <n>    fake simulation to satisfy justice property 'j<n>'\n"
     "\n"
-    "  --states  print all states\n"
+    "  --states                 print all states\n"
+    "  --vcd <file>             write VCD trace to <file>\n"
+    "  --hierarchical-symbols   interpret '.' in symbol names as hierarchical\n"
+    "                           module path in VCD\n"
+    "  --info <file>            read additional information for clocks and top\n"
+    "                           module name from <file>\n"
```
[the remark this answers is not in the document]
Comment (nakengelhardt) [reply src/btorsim/btorsim.cpp:57]: > The `--info` option allows specifying a file that contains additional information for the VCD file. The format is whitespace separated key-value, one line per pair. Key values are:
>     * `name <string>`: sets the name of the top module in the VCD.
>     * `(posedge|negedge|event) <int>`: treats the state with the given ID as a clock of the given polarity, in the first two cases introducing value changes inbetween simulation steps, and for `event` changing the type of VCD signal.
>     * `bad <int>`: treat the state with the given ID as a bad state and check that it stays false during simulation. (For use in creating cover traces, as discussed [here](https://github.com/Boolector/btor2tools/pull/4#discussion_r405573892))

It's a bit of a long description to put in the usage string, should I include it there anyway? I've added it to the comments on the reader function in `btorsimvcd.h` for now...

**Comment by mpreiner** [comment]
Merged with c35cf1c

**Pull request closed without merging**